src/logic/validateProject.js

```diff
@@ -0,0 +1,79 @@
+import map from 'lodash-es/map';
+import {createLogic} from 'redux-logic';
+
+import {validatedSource} from '../actions/errors';
+import Analyzer from '../analyzers';
+import {getCurrentProject} from '../selectors';
+import retryingFailedImports from '../util/retryingFailedImports';
+
+function importValidations() {
+  return retryingFailedImports(() =>
+    import(
+      /* webpackChunkName: "mainAsync" */
+      '../validations'
+    ),
+  );
+}
+
+async function validateSource({language, source, projectAttributes}, dispatch) {
+  const validations = await importValidations();
+  const validate = validations[language];
+  const validationErrors = await validate(source, projectAttributes);
+  dispatch(validatedSource(language, validationErrors));
+}
+
+async function validateSources({sources, projectAttributes}, dispatch) {
+  const validatePromises = map(Reflect.ownKeys(sources), language =>
```

Comment:
Definitely not going to block on this, but a minor improvement here would be to map over `sources` itself, which would give you `(source, language)` as the arguments to the iterator.